### ISSUE TITLE
Hack (ugly) to pass LD_LIBRARY_PATH to notebook

### ISSUE THREAD

**Issue by ryanmartinneutrino** [jupyterhub/jupyterhub#2630]
The issue is that I can't pass LD_LIBRARY_PATH to the containers spawned by dockerspawner.SystemUserSpawner, and have tried many of the suggestions. I'm using a container that I built from the docker-stacks. 

I need LD_LIBRARY_PATH set to a location in the container that I built (used by CERN's pyROOT)

I currently have the following in jupyterhub config:
```
c.Spawner.environment = {
   'LD_LIBRARY_PATH' : '/usr/local/software/quadis/latest/quadis/build/lib/:/opt/conda/envs/python2/lib/' ,
   'LIPA' : '/usr/local/software/quadis/latest/quadis/build/lib/:/opt/conda/envs/python2/lib/' ,
   'PYTHONPATH': '/opt/conda/envs/python2/lib:/opt/conda/envs/python2/bin/',
   }
```
I've been checking if the env variables are set by simply opening a terminal from the Jupyter Notebook home page and echo-ing, e.g $LD_LIBRARY_PATH (blank, not set), and $PYTHONPATH and $LIPA which both are set correctly. 

It sounds like LD_LIBRARY_PATH is a special environment variable, where one needs to jump through some hoops... I've tried adding LD_LIBRARY_PATH to c.Spawner.env_keep (but that's not really what I want; namely I need to set my own value based on what's installed in the container, not what's on the system). 

Upon looking at start.sh (from base-notebook), I noted that PYTHONPATH is handled in a special way, so I did the same for LD_LIBRARY_PATH:

Line 112 of start.sh (the root version of the command) was changed to:
```
exec sudo -E -H -u $NB_USER PATH=$PATH XDG_CACHE_HOME=/home/$NB_USER/.cache PYTHONPATH=${PYTHONPATH:-} LD_LIBRARY_PATH=${LD_LIBRARY_PATH:-} "${cmd[@]}"
```
which made it work for me. I did not need to change Line 156 (the jovyan version of the command). I'm confused though, since I thought that I would be running the Line 156 as I am not running as root; my dockerfile switches back to the NB user:
```
FROM jupyter/datascience-notebook:1e374527e15a
(...)
Install stuff as root, etc.
(...)
USER root
COPY start.sh /usr/local/bin/
USER $NB_USER
```

Perhaps, these changes to start.sh could be pushed into the base version of start.sh, as they shouldn't change anything. It's an ugly hack on my side, as I can't just build from the base images and have to hack the start.sh script...

**Comment by ryanlovett** [comment]
Would it be easier to drop your additional paths into a file in /etc/ld.so.conf.d/ ?

**Comment by manics** [comment]
Systemuserspawner always starts as root:
https://github.com/jupyterhub/dockerspawner/blob/522a63fcc5706e9fce5ff759b6cccb9ae1afe312/dockerspawner/systemuserspawner.py#L144 which is why the sudo command is run.

`sudo -E` should preserve env-vars but some including `LD_LIBRARY_PATH` are special: https://serverfault.com/a/371646

**Comment by ryanmartinneutrino** [comment]
Yeah, I looked at the ld.so.conf.d/ solution, but it requires ldconfig to be run as root, which (I think) still requires hacking the start up script. I would like to set the env variable in the Jupyterhub config file, so I still need to run ldconfig as root before the notebook server starts. 

Ah, thanks for pointing that out, it was a little hard to follow which branch of start.sh systemuserspawner was going to call.

**Comment by ryanlovett** [comment]
> but it requires ldconfig to be run as root

Why is that a barrier though since you can set USER?

**Comment by ryanmartinneutrino** [comment]
I think the issue is when to run ldconfig, since it has to be run after the container build process (as I need to set LD_LIB from the jupyterhub host). My understanding is that I have to stick that command into one of the start up scripts so that it runs before the notebook is started.

**Comment by manics** [comment]
Can you add the path to `/etc/ld.so.conf.d` at build time in the Dockerfile, and run ldconfig also in the Dockerfile, instead of at runtime?

**Comment by ryanmartinneutrino** [comment]
I have to set the value of LD_LIB in the jupyterhub config file, so ldconfig has to run after the docker image is built in my use case. It's admittedly a weird use case, as LD_LIB is not pointing to something I installed in the Dockerfile, but rather to a library on the host (mounted as a volume), whose location may change (and I don't want to recompile the notebook image each time it does).

**Comment by manics** [comment]
I'm closing this as it's quite old and it's a support issue. If you're still having problems please post on the [Jupyter Community Forum](https://discourse.jupyter.org/). Thanks!

**Comment by fxheng** [comment]
Hi, 
Just went through the problem of setting a LD_LIBRARY_PATH available in the lab and get a simple workaround at container build time. While in ROOT just add your env variable in /etc/environment.

`RUN echo LD_LIBRARY_PATH="/usr/local/lib:/usr/lib/x86_64-linux-gnu/odbc/" >> /etc/environment `

**Comment by yuyouling** [comment]
The solution is very simple. 
1. use **jupyterhub --generate-config** to generate the **jupyterhub_config.py**
2. use text editor such as **nano** or **vim** to open the **jupyterhub_config.py**
3. change some configs
(1)  general setting
      c.LocalAuthenticator.create_system_users = True
      c.PAMAuthenticator.encoding = 'utf8'
      c.PAMAuthenticator.open_sessions = False
      c.JupyterHub.admin_access = True
      c.JupyterHub.ip = '0.0.0.0'
      c.JupyterHub.port = 8888
      c.Spawner.args = ['--allow-root']
      c.Spawner.default_url = '/lab'
      c.Spawner.notebook_dir = '~'
      c.Authenticator.admin_users = set(['root', 'alex'])
(2) LD_LIBRARY_PATH related setting
      c.Spawner.env_keep = ['LD_LIBRARY_PATH']       #use for keep the environment variable - LD_LIBRARY_PATH
      c.Spawner.cmd = ['/opt/anaconda3/envs/pytorch2/bin/jupyterhub-singleuser']        # !**!! very important**, you should use the full path of the command **jupyterhub-singleuser**, due to the Spawner process the activate the **jupyterhub-singleuser** the load the environment (I think it does as this way)
(3) save this file
4. run **jupyterhub -f jupyterhub_config.py** to start jupyterhub.  
5. open the jupyterlab website and create the notebook file, try the following scripts, you can see the expected result.
     import os
     os.environ['LD_LIBRARY_PATH']
